### ISSUE TITLE
Creating a GC for a device paint "somewhere" on the screen on windows

### ISSUE THREAD

**Issue by laeubi** [eclipse-platform/eclipse.platform.swt#198]
This was originally reported as [Bug 572400](https://bugs.eclipse.org/bugs/show_bug.cgi?id=572400):

If I create a GC in windows for a device, paint on it and dispose it afterward the painted items appear "somewhere" on the screen. Some of the painting even persists the termination of the application. Under Linux everything works fine.

I have [attached a snippet](https://bugs.eclipse.org/bugs/attachment.cgi?id=285977) to demonstrate the problem:

- start the snippet and move the mouse --> application is covered with black parts even outside the main window
- move the window, the black parts appear on other parts of the desktop
- terminate the application --> some of the parts remain and require to force a repaint (e.g. minimize other application windows and restore afterwards)

### running application:
![](https://bugzillaattachments.eclipsecontent.org/bugs/attachment.cgi?id=285978)

### terminated application:
![](https://bugzillaattachments.eclipsecontent.org/bugs/attachment.cgi?id=285979)

We see strange GC behavior on SWT windows as well on other (not yet reproducible) cases, so it would be good if someone with windows knowledge could take a look at this case here, maybe this reveals some other issues as well.

**Comment by Phillipus** [comment]
It depends on the drawable declared in `GC`.

That should be:

`GC gc = new GC(canvas);`

not

`GC gc = new GC(canvas.getDisplay());`

**Comment by laeubi** [comment]
@Phillipus I don't want to draw on the canvas (I could better use a paint listener for this)!

**Comment by SyntevoAlex** [comment]
In Windows, it is valid to paint to screen directly. I myself used it a few times (for example, to output debug infos). It is expected that affected windows remain "dirty" until they are force painted (because they are not invalidated after being corrupted by you).

Why are you painting to screen? What did you actually want to do?

**Comment by laeubi** [comment]
> In Windows, it is valid to paint to screen directly.

Is this documented in SWT somewhere, this was rather unexpected behavior.

> It is expected that affected windows remain "dirty" until they are force painted (because they are not invalidated after being corrupted by you).

This really seems dangerous to me if I can paint over others windows from an arbitrary SWT application.

> Why are you painting to screen?

This was used to "simulate" graphic drawing without allocating an image, as mentioned on Linux (and macOS) this drawing is just invisible and it was rather disturbing to see "magic" things appearing on the screen and even not getting cleared at all.

**Comment by Phillipus** [comment]
I can understand it happening as it does on Windows, but don't understand why it doesn't paint anything on Linux and macOS.

**Comment by SyntevoAlex** [comment]
Probably Linux and macOS do not allow to paint to screen directly?

**Comment by laeubi** [comment]
> but don't understand why it doesn't paint anything on Linux and macOS

Maybe this is just a bug on Linux and macOS then? :-)

**Comment by Phillipus** [comment]
> Probably Linux and macOS do not allow to paint to screen directly?

But it won't even paint on the Canvas object by using `GC gc = new GC(canvas);`

**Comment by Phillipus** [comment]
> Try painting in a timer.

Nothing painted using timer.

**Comment by SyntevoAlex** [comment]
That's a bit unexpected, but I'm lazy to investigate it. Maybe Linux and macOS forbid painting outside `SWT.Paint` event, or maybe you have a mistake in your code.

**Comment by SyntevoAlex** [comment]
> This was used to "simulate" graphic drawing without allocating an image

I still don't get it, what's the point in drawing that has no expected effect?

**Comment by laeubi** [comment]
> > This was used to "simulate" graphic drawing without allocating an image
> 
> I still don't get it, what's the point in drawing that has no expected effect?

@SyntevoAlex I don't think it is that relevant here as I just created the example to show a certain behavior that differs between SWT platforms.

As it is different on Windows, I assumed it is a bug on Windows, but maybe I'm wrong and it is actually a bug that Linux/Mac behave different, it is just a bit confusing if it "works" and as soon as you run your application on windows produces strange results.

**Comment by niraj-modi** [comment]
> It depends on the drawable declared in `GC`.
> 
> That should be:
> 
> `GC gc = new GC(canvas);`
On Windows10 drawing is limited to canvas:
- This seems expected.
> 
> not
> 
> `GC gc = new GC(canvas.getDisplay());`
On Windows10 drawing is applicable to complete display:
- It's kind of expected as we are getting what we coded for.

> @SyntevoAlex I don't think it is that relevant here as I just created the example to show a certain behavior that differs between SWT platforms.
> 
> As it is different on Windows, I assumed it is a bug on Windows, but maybe I'm wrong and it is actually a bug that Linux/Mac behave different, it is just a bit confusing if it "works" and as soon as you run your application on windows produces strange results.

@laeubi : IMO let's re-purpose this bug for investigating Linux/MAC behavior. WDYT ?
